src/engine/TPA.cc

```diff
@@ -1379,98 +1413,115 @@ void TransitionSystemNetworkManager::initNetwork() {
             outGoingEdges.erase(std::remove_if(outGoingEdges.begin(), outGoingEdges.end(), [this](EId eid) {
                 return graph.getSource(eid) == graph.getTarget(eid);
             }), outGoingEdges.end());
-            assert(outGoingEdges.size() == 1);
-            EId edge = outGoingEdges[0];
-            if (graph.getTarget(edge) != graph.getExit()) {
-                getNode(graph.getTarget(edge)).previous = edge;
+            for (auto edge: outGoingEdges) {
+                if (graph.getTarget(edge) != graph.getExit()) {
+                    getNode(graph.getTarget(edge)).parents.push(edge);
+                }
             }
         }
         if (vid != graph.getEntry()) {
             auto incomingEdges = adjacencyRepresentation.getIncomingEdgesFor(vid);
             incomingEdges.erase(std::remove_if(incomingEdges.begin(), incomingEdges.end(), [this](EId eid) {
                 return graph.getSource(eid) == graph.getTarget(eid);
             }), incomingEdges.end());
-            assert(incomingEdges.size() == 1);
-            EId edge = incomingEdges[0];
-            if (graph.getSource(edge) != graph.getEntry()) {
-                getNode(graph.getSource(edge)).next = edge;
+            for (auto edge: incomingEdges) {
+                if (graph.getSource(edge) != graph.getEntry()) {
+                    getNode(graph.getSource(edge)).children.push(edge);
+                }
             }
         }
     }
 }
 
 VerificationResult TransitionSystemNetworkManager::solve() && {
     initNetwork();
-
-    auto current = getChainStart();
-    while (true) {
-        auto [res,explanation] = queryTransitionSystem(networkMap.at(current));
-        if (reachable(res)) {
-            if (current == getChainEnd()) {
-                return {VerificationAnswer::UNSAFE, computeInvalidityWitness()};
-            }
-            EId nextEdge = getOutgoingEdge(current);
-//            std::cout << "Reached states in node " << current.id << ": " << logic.pp(explanation) << std::endl;
-            getNode(current).trulyReached = explanation;
-            PTRef nextConditions = getInitialStates(graph.getTarget(nextEdge));
-            auto [edgeRes, edgeExplanation] = queryEdge(nextEdge, explanation, nextConditions);
-            if (reachable(edgeRes)) { // Edge propagates forward
-                auto next = graph.getTarget(nextEdge);
-                networkMap.at(next).solver->resetInitialStates(edgeExplanation);
-                current = next;
-//                std::cout << "Propagating forward query to " << next.id << ": " << logic.pp(edgeExplanation) << std::endl;
-                continue; // Information has been propagated to the next node, continue querying that node
-
-            } else { // Edge cannot propagate forward
+    auto graphStarts = getGraphStarts();
+    auto graphEnds = getGraphEnds();
+    bool safe;
+    for(auto init: graphStarts){
```

Comment:
Instead of iterating over graphStarts, you can just iterate over outgoing edges from CHCGraph entry.